src/protocols/aggsig/musig_two_rounds.rs

```diff
@@ -0,0 +1,376 @@
+            })
+            .collect();
+
+        let pk1 = X_tilde_vec.remove(0);
+        let sum = X_tilde_vec
+            .iter()
+            .fold(pk1, |acc, pk| acc.add_point(&pk.get_element()));
+
+        KeyAgg {
+            X_tilde: sum,
+            a_i: hash_vec[party_index].clone(),
+        }
+    }
+}
+
+#[derive(Debug,Clone)]
+pub struct EphemeralKey {
+    pub keypair: KeyPair,
+    pub commitment: BigInt,
+    pub blind_factor: BigInt,
+}
+
+
+
+
+impl EphemeralKey {
+    pub fn create_vec_from_private_key(x1: &KeyPair) -> Vec<EphemeralKey> {
+        let mut EphermalKeys_vec: Vec<EphemeralKey> = vec![];
+        for i in 0..Nv {
+            let base_point: GE = ECPoint::generator();
+            let hash_private_key_message =
+                HSha256::create_hash(&[&x1.private_key.to_big_int(), &BigInt::from(i as i32)]);
+            let ephemeral_private_key: FE = ECScalar::from(&hash_private_key_message);
+            let ephemeral_public_key = base_point.scalar_mul(&ephemeral_private_key.get_element());
+            let (commitment, blind_factor) = HashCommitment::create_commitment(
+                &ephemeral_public_key.bytes_compressed_to_big_int(),
+            );
+            let eph_key = EphemeralKey {
+                keypair: KeyPair {
+                    public_key: ephemeral_public_key,
+                    private_key: ephemeral_private_key,
+                },
+                commitment,
+                blind_factor,
+            };
+            EphermalKeys_vec.push(eph_key);
+        }
+        EphermalKeys_vec
+     }
+
+}
+
+
+
+#[derive(Debug, Clone)]
+pub struct State {
+    pub keypair: KeyPair,
+    pub ephk_vec: Vec<EphemeralKey>,
+}
+
+
+
+
+impl State {
+    fn add_ephemeral_keys(&self, msg_vec: &[Vec<GE>]) -> Vec<GE> {
+        let mut R_j_vec: Vec<GE> = vec![];
+        for j in 0..Nv {
+            let pk_0j = self.ephk_vec[j].keypair.public_key;
+            let R_j: GE = msg_vec.
+                iter().
+                fold(pk_0j, |acc, ephk| acc.add_point(&ephk.get(j).unwrap().get_element()));
+            R_j_vec.push(R_j);
+        }
+        R_j_vec
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct StatePrime {
+    pub R: GE,
+    pub s_i: FE,
+}
+
+pub fn sign(x: KeyPair) -> ( Vec<GE>, State) {
+    let ephk_vec = EphemeralKey::create_vec_from_private_key(&x);
+    let msg = ephk_vec
+        .iter()
+        .map(|eph_key| eph_key.keypair.public_key)
+        .collect();
+    (msg, State {keypair: x, ephk_vec: ephk_vec})
+}
+
+
+pub fn hash_sig(r_hat: &GE, X_tilde: &GE, message: &[u8]) -> BigInt {
+    HSha256::create_hash(&[
+        &BigInt::from(0),
+        &r_hat.x_coor().unwrap(),
+        &X_tilde.bytes_compressed_to_big_int(),
+        &BigInt::from(message),
+    ])
+}
+
+impl StatePrime {
+    pub fn get_StatePrime(&self) -> &StatePrime {
+        self
+    }
+}
+
+     fn compute_signature_share(
+         state: &State,
```

Comment:
and also, it looks like that `compute_signature_share`, `compute_global_params` and `sign_prime` 
should be put inside 
```
impl State {
...
}
```
instead of 
```
impl StatePrime {
...
}
```

Reply:
Currently nothing implements StatePrime, but I am considering whether or not to put it inside impl State